---
title: "astrocytes_malevsfemale"
author: "Lucy"
date: "2023-12-04"
output: html_document
---

---
title: "229E_astro_Infection_differential_expression"
author: "Lucy"
date: "2023-10-01"
output: html_document
---


### Load librarys
```{r}
# install.packages("datawizard")

library(readr)
library(knitr)
library(tidyverse)
library(GEOquery)
library(limma)
library(statmod)
library(ggplot2)
library(datawizard)
library(reshape2)
library(ggplot2)
library(ggfortify)
library(ggrepel)
library(tidyverse)
library(fgsea)
library(ggfortify)
library(msigdbr) 
library(data.table)
library(R.utils)
library(enrichplot)
library(DOSE)
library(clusterProfiler)
library(ggridges)
library(ReactomePA)
library(org.Hs.eg.db)
library(biomaRt)
```

###import data
```{r}
#protein data
astro_raw_data <- read_tsv(file = '../data/combined_protein_astrocytes_sex_time.tsv')

nrow(astro_raw_data)
#meta data
astro_meta_data <- read_tsv(file = '../data/experiment_annotation_astrocytes_sex_time.tsv')

astro_meta_data$sex <- c("F", "F", "F", "F", "F", "F", "M", "M", "M", "M", "M", "M")
astro_meta_data$time <- c("T33", "T33", "T33", "T37", "T37", "T37", "T33", "T33", "T33", "T37", "T37", "T37")
female_meta_data <- astro_meta_data[1:6, ]
male_meta_data <- astro_meta_data[7:12, ]
T37_meta_data <- astro_meta_data[c(4:6, 10:12), ]
```

###pre-process/cleaning data:
####1. remove contamination
####2. remove unnecessary columns
####3. log transformation if required
####4. remove lowly expressed proteins
```{r}
#1. remove contamination
raw_data_wo_contam <- astro_raw_data[- grep("contam_sp", astro_raw_data$Protein),]

#2. remove unnecessary columns
shorten_data <- select(raw_data_wo_contam, Protein, "Protein ID", "Entry Name", Description, contains("MaxLFQ Intensity"))
shorten_data[shorten_data == 0] <- NA



#3. log transformation

#this example shows that the data requires log transformation (not a bell shaped curve). 
# six_hpi1 <- hist(infec_data_shorten$X215_LucyC_6hpi1_100ng_1.Total.Spectral.Count,
# main="Protein Expression Histogram",
# xlab="Expression",
# xlim=c(0,100),
# col="darkmagenta",
# breaks = 1000)

numeric_MaxLFQ_data <- shorten_data %>% select(contains("MaxLFQ Intensity"))
rownames(numeric_MaxLFQ_data) <- shorten_data$Protein
logged_data <- log2(numeric_MaxLFQ_data)
rownames(logged_data) <- shorten_data$Protein

female_logged_data <- select(logged_data, contains("female"))
female_logged_data$Description = shorten_data$Description
#column_to_rownames(female_logged_data, "Description")
rownames(female_logged_data) <- shorten_data$Protein

male_logged_data <- logged_data[ , 7:12]
rownames(male_logged_data) <- shorten_data$Protein

T37_logged_data <- logged_data[ , c(4:6, 10:12)]
rownames(T37_logged_data) <- shorten_data$Protein

# #4. remove lowly expressed proteins
# #convert "-inf" to "NA"
# infec_data_logged[infec_data_logged == "-Inf"] <- NA
# 
# #make a new column counting the number of "NA" values
# count <- rowSums(is.na(infec_data_logged))
# data.log.filt <- data.frame(cbind(infec_data_logged, "Count" = count))
# 
# #Adjust the threshold at your discretion. 
# #Here, I am allowing up to 75% of protein observations to be "NA"
# data.log.filt <- subset(data.log.filt, Count < ncol(infec_data_logged)*0.75)
# data.log.filt <- subset(data.log.filt, select = -c(Count)) #Remove count variable
```

## Without Log transform (For PCA)
```{r}
count <- rowSums(is.na(numeric_MaxLFQ_data))
data.filt <- data.frame(cbind(numeric_MaxLFQ_data, "Count" = count))

data.filt <- subset(data.filt, Count < ncol(numeric_MaxLFQ_data)*0.75)
data.filt <- subset(data.filt, select = -c(Count))

```

```{r}

data.filt.pca <- t(data.filt) %>% as.data.frame()

## replace NA with 0 because that should mean 0 spectral count
data.filt.pca[is.na(data.filt.pca)] <- 0

## A way you can check which column has all the same values (if any)
# which(apply(data.filt.pca, 2, var)==0)

# data.filt.pca = data.filt.pca %>% select(-c(rowname))

data.filt.pca %>% head()

p <- prcomp(data.filt.pca, scale = TRUE)

data.filt.pca = rownames_to_column(data.filt.pca)
data.filt.pca$condition = c("F1_33C", "F2_33C", "F3_33C", "F1_37C", "F2_37C", "F3_37C", "M1_33C", "M2_33C", "M3_33C", "M1_37C", "M2_37C", "M3_37C")
#data.filt.pca$hpi = data.filt.pca$hpi %>%  as.character()   # this makes the colour scale not an idiot

autoplot(p, data=data.filt.pca, colour = "condition")

```


###quality control
```{r}
#histogram

# first, transpose data to standardize by protein
# (by default, the standardize() function will calculate z scores by column,
# so we want our proteins as columns now
data.t <- data.frame(t(data.filt))
data.t.z <- standardize(data.t)

# Now, we need the data in long-form (melted) for making the heatmap.
# we want to bind the sample metadata to this matrix ahead of melting
# First make sure the protein data and the metadata are ordered the same
rownames(data.t.z)==astro_meta_data$sample #check

# The order is not the same. Order both objects and try again
meta_data.order <- meta_data[order(meta_data$sample),]

#I am sure you could order the rownames directly too, but here I move the sample names into a column, order it, and move it back to rownames
data.t.z <- rownames_to_column(data.t.z) 
data.t.z <- data.t.z[order(data.t.z$rowname),]
rownames(data.t.z) <- NULL #uninitialize rownames to accept new names
data.t.z <- column_to_rownames(data.t.z, var = "rowname")
rownames(data.t.z) == astro_meta_data$sample #Success!

#now that it's ordered, bind metadata variables to protein data
data.t.z <- data.frame(cbind(data.t.z, astro_meta_data))

#define vector for id.vars
vars <- colnames(astro_meta_data)
data.t.z.melt <- melt(data.t.z, id.vars = vars, value.name = "Z.score", variable.name = "Protein")
head(data.t.z.melt)

ggplot(data.t.z.melt, aes(x = condition, y = Protein, fill = Z.score)) +
  geom_tile()+
  scale_fill_gradient2(low = "#0571b0", mid = "white", high = "#ca0020",
                       name = "Z score", limits = c(-4,4),na.value = "grey")+
  theme_bw()+
  theme(axis.text.y = element_blank(),
        axis.text.x = element_text(angle = -50, hjust = 0, size = 6, color = "black"),
        plot.background = element_blank(),
        axis.title.x = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        strip.background = element_blank(),
        axis.ticks.y = element_blank(),
        legend.position = "top")+
  scale_y_discrete(expand = c(0,0)) + 
  scale_x_discrete(expand = c(0,0))  

#data look uniform without major outliers

```

```{r}
# # check for data quality with scattered plots
# 
# histogram_log_data_sample <- data.log.filt %>%  select("X215_LucyC_6hpi1_100ng_1.Total.Intensity")
# 
# histogram_log_data_sample %>% head()
# 
# ggplot(histogram_log_data_sample, aes(x=X215_LucyC_6hpi1_100ng_1.Total.Intensity)) + geom_histogram()
# 
# histogram_log_data <- data.frame(t(data.log.filt)) %>%  select(0:1)
# 
# head(histogram_log_data)
# 
# ggplot(histogram_log_data, aes(x=sp.A0A0B4J2D5.GAL3B_HUMAN)) + geom_histogram()
# 
# library(GGally)
# 
# 
# data.filt
# ggpairs(data.log.filt, columns = c(1, 4, 6))
# ggpairs(data.log.filt, columns = c(4, 5, 6))

```

###limma differential analysis
```{r}
# colnames(data.filt) == meta_data.order$sample #if false, order them
#colnames(data.log.filt) == astro_meta_data$sample #if false, order them


#Call limma - male astrocytes at 33C and 37C
male_desMat <- model.matrix(~ time, male_meta_data)

male_fit <- lmFit(male_logged_data, male_desMat)
colnames(male_fit)

male_tmp <- eBayes(male_fit)

male_sig.prots <- topTable(male_tmp, coef = "timeT37", n = Inf, p.value = 0.05)
#male_sig.prots <- subset(limma.results, adj.P.Val < 0.05)
head(male_sig.prots)


#Call limma - female astrocytes at 33C and 37C
female_desMat <- model.matrix(~ time, female_meta_data)

female_fit <- lmFit(female_logged_data, female_desMat)
colnames(female_fit)

female_tmp <- eBayes(female_fit)

female_sig.prots <- topTable(female_tmp, coef = "timeT37", n = Inf, p.value = 0.05)
#female_sig.prots <- subset(limma.results, adj.P.Val < 0.05)
head(female_sig.prots)

#Call limma - T37 astrocytes for male and female
T37_desMat <- model.matrix(~ sex, T37_meta_data)

T37_fit <- lmFit(T37_logged_data, T37_desMat)
colnames(T37_fit)

T37_tmp <- eBayes(T37_fit)

T37_sig.prots <- topTable(T37_tmp, coef = "sexM", n = Inf, p.value = 0.05)
T37_prots <- topTable(T37_tmp, coef = "sexM", n = Inf)
head(T37_sig.prots)

#interaction
#Call limma - T37 astrocytes for male and female
int_desMat <- model.matrix(~ sex*time, astro_meta_data)

int_fit <- lmFit(logged_data, int_desMat)
colnames(int_fit)

int_tmp <- eBayes(int_fit)

int_sig.prots <- topTable(int_tmp, coef = "sexM:timeT37", n = Inf)
head(T37_sig.prots)

#Call limma - T37 astrocytes for male and female
add_desMat <- model.matrix(~ sex+time, astro_meta_data)

add_fit <- lmFit(logged_data, add_desMat)
colnames(add_fit)

add_tmp <- eBayes(add_fit)

add_sig.prots <- topTable(int_tmp, coef = "sexM", n = Inf, p.value = 0.05)
```


```{r}
# The significantly differentially expressed genes are the ones found in the upper-left and upper-right corners.
# Add a column to the data frame to specify if they are UP- or DOWN- regulated (log2FoldChange respectively positive or negative)


## Split the Protein name by delimiter "|"
T37_prots <- separate_wider_delim(T37_prots, cols=Protein, delim = "|", names = c("first", "second", "third"))
T37_prots <- separate_wider_delim(T37_prots, cols = third, delim = "_", names= c("name", "human"))
T37_prots <- T37_prots %>% select(-c("first", "second", "human"))

# add a column of NAs
T37_prots$diffexpressed <- "NO"
# if log2Foldchange > 0.6 and pvalue < 0.05, set as "UP" 
T37_prots$diffexpressed[T37_prots$logFC > 0 & T37_prots$adj.P.Val < 0.05] <- "UP"
# if log2Foldchange < -0.6 and pvalue < 0.05, set as "DOWN"
T37_prots$diffexpressed[T37_prots$logFC < 0 & T37_prots$adj.P.Val < 0.05] <- "DOWN"

# protein labels
# T37_prots <- rownames_to_column(T37_prots, "Protein")

T37_prots$delabel <- NA
T37_prots$delabel[T37_prots$diffexpressed != "NO"] <- T37_prots$name[T37_prots$diffexpressed != "NO"]

# Re-plot but this time color the points with "diffexpressed"
ggplot(data=T37_prots, aes(x=logFC, y=-log10(adj.P.Val), col=diffexpressed, label=delabel)) + geom_point() + theme_minimal() + geom_text_repel(size = 3, colour = "black") +
         scale_color_manual(values=c("pink", "gray", "plum3")) +
        geom_vline(xintercept=0, col="gray20") +
        geom_hline(yintercept=-log10(0.05), col="gray20")


```

```{r}
knitr::kable(msigdbr::msigdbr_species())

#Access the KEGG gene sets. 
kegg.human.db <- msigdbr::msigdbr(species = "human",
                                  category = "C2",
                                  subcategory = "CP:KEGG")
IMMUNE.human.db <- msigdbr(species = "human", 
                           category = "C7", 
                           subcategory = "IMMUNESIGDB")
```

```{r}
# obtain GWAS catalog information

run.if.needed <- function(.file, .code) {
    if(!file.exists(.file)) { .code }
    stopifnot(file.exists(.file))
}

gwas.tidy.file <- "gwas_catalog_tidy.tsv.gz"

run.if.needed(gwas.tidy.file, {
    gwas.file <- "gwas_catalog_v1.0-associations_e105_r2022-02-02.tsv.gz"
    run.if.needed(gwas.file, {
        url <- "https://www.ebi.ac.uk/gwas/api/search/downloads/full"
        .file <- str_remove(gwas.file, ".gz$")
        download.file(url, destfile = .file)
        gzip(.file)
        unlink(.file)
    })
    .dt <-
        fread(gwas.file, sep="\t", quote="") %>%
        dplyr::select(`MAPPED_GENE`, `DISEASE/TRAIT`, `PVALUE_MLOG`)
    .dt <- .dt[order(.dt$PVALUE_MLOG, decreasing = TRUE),
               head(.SD, 1),
               by = .(`MAPPED_GENE`, `DISEASE/TRAIT`)]
    .count <- .dt[, .(.N), by = .(`DISEASE/TRAIT`)]
    .dt <- left_join(.count[`N` >= 100, ], .dt)[nchar(`MAPPED_GENE`)> 0,]
    .dt <- .dt[,
               .(gene_symbol = unlist(strsplit(`MAPPED_GENE`, split="[ ,.-]+"))),
               by = .(`DISEASE/TRAIT`, PVALUE_MLOG)]
    .dt[, p.value := 10^(-PVALUE_MLOG)]

    fwrite(.dt, file=gwas.tidy.file)
})
```

